### PR TITLE
Make var(Range) type stable

### LINKED ISSUE #22773: Implementation of var(v::Range) is not type-stable
Return type of, e.g., `var(0f0:1f0:10f0)` is `Union{Float32, Float64}`, due to conditional return of `NaN` in the implementation.

Example:

       @code_warntype var(0f0:1f0:10f0)q

`var(v::Range{T})` should probably return `real(momenttype(T))` in all cases, like `var(A::AbstractArray{T})` does.

May also affect implementation of similar functions.

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes #22773

### CODE REVIEW

File: base/statistics.jl [hunk 1]
```diff
@@ -210,22 +210,24 @@ varm(iterable, m::Number; corrected::Bool=true) =
 ## variances over ranges
 
 function varm(v::Range, m::Number)
-    f = first(v) - m
-    s = step(v)
-    l = length(v)
+    f  = first(v) - m
+    s  = step(v)
+    l  = length(v)
+    vv = f^2 * l / (l - 1) + f * s * l + s^2 * l * (2 * l - 1) / 6
```
Comment (tkelman) [inline base/statistics.jl]: wouldn't this throw for length 1 integer ranges?

Comment (andreasnoack) [inline base/statistics.jl]: No, because `/` promotes to floating point.

File: test/statistics.jl [hunk 2]
```diff
@@ -113,6 +113,18 @@ X = [2 3 1 -1; 7 4 5 -4]
 @test isnan(var(1:1))
 @test isnan(var(1:-1))
 
+@test @inferred(var(1.0:8.0)) == 6.
+@test varm(1.0:8.0,1.0) == varm(collect(1.0:8.0),1)
+@test isnan(varm(1.0:1.0,1.0))
+@test isnan(var(1.0:1.0))
+@test isnan(var(1.0:-1.0))
+
+@test @inferred(var(1.0f0:8.0f0)) == 6.f0
```
Comment (tkelman) [inline test/statistics.jl]: `===`

File: test/statistics.jl [hunk 3]
```diff
@@ -113,6 +113,18 @@ X = [2 3 1 -1; 7 4 5 -4]
 @test isnan(var(1:1))
 @test isnan(var(1:-1))
 
+@test @inferred(var(1.0:8.0)) == 6.
+@test varm(1.0:8.0,1.0) == varm(collect(1.0:8.0),1)
```
Comment (Sacha0) [inline test/statistics.jl]: Spaces following commas in the argument lists in this and similar lines below might improve readability :).

File: test/statistics.jl (hunk 2 above)
Comment (andreasnoack) [inline test/statistics.jl]: What would be the difference when testing against `6.0f0`?

File: test/statistics.jl (hunk 3 above)
Comment (andreasnoack) [inline test/statistics.jl]: I agree but the style in this file seems to be compact so I'll leave to a separate PR to change that.

File: test/statistics.jl (hunk 2 above)
Comment (tkelman) [inline test/statistics.jl]: it would check that the return type is Float32

**Pull request merged**